torch/nn/parallel/distributed.py

```diff
@@ -0,0 +1,313 @@
+import sys
+import math
+import threading
+
+import torch
+from torch import nn
+from torch.autograd import Variable
+from torch._utils import _flatten_tensors, _unflatten_tensors
+from torch.cuda.comm import broadcast_coalesced
+from torch.cuda import nccl
+import torch.distributed as dist
+
+from .replicate import replicate
+from .scatter_gather import scatter_kwargs, gather
+from .parallel_apply import parallel_apply
+
+if sys.version_info[0] == 3:
+    import queue
+else:
+    import Queue as queue
+
+
+class DistributedDataParallel(nn.Module):
+    """Implements distributed data parallelism at the module level.
+
+    This container parallelizes the application of the given module by
+    splitting the input across the specified devices by chunking in the batch
+    dimension. The module is replicated on each machine and each device, and
+    each such replica handles a portion of the input. During the backwards
+    pass, gradients from each node are averaged.
+
+    The batch size should be larger than the number of GPUs used locally. It should
+    also be an integer multiple of the number of GPUs so that each chunk is the
+    same size (so that each GPU processes the same number of samples).
+
+    See also: :ref:`cuda-nn-dataparallel-instead`. The same constraints on input
+    as in :class:`torch.nn.DataParallel` apply.
+
+    .. warning::
+        Constructor, forward method, and differentiation of the output (or a
+        function of the output of this module) is a distributed synchronization
+        point. Take that into account in case different processes might be executing
+        different code.
+
+    .. warning::
+        This module assumes all parameters are registered in the model by the time
+        it is created. No parameters should be added nor removed later. Same applies
+        to buffers.
+
+    .. warning::
+        This module doesn't work with :func:`torch.autograd.grad` (i.e. it will only
+        work if gradients are to be accumulated in ``.grad`` attributes of parameters).
+
+    .. note::
+        Parameters are never broadcast between processes. The module performs
+        an all-reduce step on gradients and assumes that they will be modified by the
+        optimizer in all processes in the same way. Buffers (e.g. BatchNorm stats) are
+        broadcast form the module in process of rank 0, to all other replicas in the
+        system in every iteration.
+
+    Args:
+        module: module to be parallelized
+        device_ids: CUDA devices (default: all devices)
+        output_device: device location of output (default: device_ids[0])
+
+    Example::
+
+        >>> net = torch.nn.DataParallel(model, device_ids=[0, 1, 2])
```

Comment:
`s/DataParallel/DistributedDataParallel`

If there are environment variables that need to be set to use it, this would be a good place give some examples of that too.

Same with any functions you need to call to rendezvous before creating DistributedDataParallel